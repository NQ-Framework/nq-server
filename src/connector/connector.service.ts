import { Injectable } from '@nestjs/common';
import { Response } from '@irreal/nestjs-sse';

@Injectable()
export class ConnectorService {
  responseMap = {};
  add(id: string, res: Response) {
    this.responseMap[id] = res;
  }
  get(id: string): Response | undefined {
    if (!(id in this.responseMap)) {
      return undefined;
    }
    return this.responseMap[id] as Response;
  }

  has(id: string): boolean {
    return this.responseMap[id] !== undefined;
  }

  remove(id: string) {
    delete this.responseMap[id];
  }

  send(id: string, data: any) {
    const res = this.get(id);
    if (res === undefined) {
      throw new Error("Can't send data to a non existing client");
    }
    res.sse(`data: ${data}\n\n`);
  }
}
