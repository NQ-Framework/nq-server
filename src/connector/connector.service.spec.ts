import { Test, TestingModule } from '@nestjs/testing';
import { ConnectorService } from './connector.service';
import { Response } from '@irreal/nestjs-sse';

describe('ConnectorService', () => {
  let service: ConnectorService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ConnectorService],
    }).compile();

    service = module.get<ConnectorService>(ConnectorService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should store and retrieve response objects', () => {
    const res: any = {
      sse: () => {
        return 'sse';
      },
    };

    const res2: any = {
      sse: () => {
        return 'sse 2';
      },
    };
    service.add('test id', res);
    service.add('test id 2', res2);

    const resReturned = (service.get('test id') as any) as Response;
    expect(resReturned.sse('')).toEqual('sse');
    expect(resReturned).toEqual(res);

    const resReturned2 = (service.get('test id 2') as any) as Response;
    expect(resReturned2).toEqual(res2);
    expect(resReturned2.sse('')).toEqual('sse 2');
  });

  it('should return undefined for missing response objects', () => {
    expect(service.get('missing id')).toBeUndefined();
  });

  it('should replace the object if added twice with same id', () => {
    service.add('test id', { sse: () => 'test 1' } as any);
    expect(((service.get('test id') as any) as Response).sse('')).toEqual(
      'test 1',
    );
    service.add('test id', { sse: () => 'test 2' } as any);
    expect(((service.get('test id') as any) as Response).sse('')).toEqual(
      'test 2',
    );
  });

  it('should check if response is present', () => {
    expect(service.has('test id')).toEqual(false);
    service.add('test id', {} as any);
    expect(service.has('test id')).toEqual(true);
    service.remove('test id');
    expect(service.has('test id')).toEqual(false);
  });

  it('should send data to response object', (done) => {
    const res: any = {
      sse: (data: any) => {
        expect(data).toEqual('data: test\n\n');
        done();
      },
      end: () => {
        // this is a hack until google cloud run enables http streaming
      },
    };
    service.add('test id', res);
    service.send('test id', 'test');
  });

  it('should throw when sending data to non existing id', () => {
    expect(() => {
      service.send('test id', 'test');
    }).toThrowErrorMatchingSnapshot();
  });

  it('should close the connection', (done) => {
    const res: any = {
      end: () => {
        done();
      },
    };
    service.add('test id', res);
    service.close('test id');
  });

  it('kasot', () => {
    expect(() => {
      service.close('test id');
    }).toThrowErrorMatchingSnapshot();
  });
});
