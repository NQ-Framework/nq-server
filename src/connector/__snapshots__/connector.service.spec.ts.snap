// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectorService should throw when sending data to non existing id 1`] = `"Can't send data to a non existing client"`;
